### ISSUE TITLE
Make illegal states unrepresentable.

### ISSUE THREAD

**Issue by jlouis** [jlouis/etorrent#78]
See [Yaron Minsky's presentation on vimeo](http://vimeo.com/14313378) for the details of this idea. We ought to replace #state{} with some more specific records.

**Comment by edwardw** [comment]
Yes, this presentation is very good. And "make illegal states unrepresentable" part is around 18:10.
